### ISSUE TITLE
PyTorch->XLA interface refinement

### ISSUE THREAD

**Issue by soumith** [pytorch/xla#4]
@dlibenzi started a discussion with:

I'd like to get the general feeling for an interface we currently use in our tests, to hide the setup complexities (tracing, and other things typical to XLA) of a traced model.

Below is how an MNIST (resnet18/cifar10 is identical, modulus network and shapes) training test code looks like.

We have an `XlaModel` class, which takes the typical PyTorch `nn.Module` the users are accustomed to, input/target shapes and types (they default to `float32`), the loss function, and the devices (num_cores and devices are optional for one core setups).

We then have two main functions, `train()` and `test()` (code should be clear enough).

One advantage of having an wrapper class like this, is that, like mentioned before, it hides the tracing and setup code, so that users won't have to repeat it all over.

With its functionality, it can be plugged under the PyTorch data parallel API AFAICT.

More importantly, it hides the details of the optimization we currently do (and which we will do in the future), opening the way to reach top TPU performance with on device loops and infeeds.
Any thoughts?
Thanks!

```python
def train_mnist():
    lr = 0.01 * FLAGS.num_cores
    momentum = 0.5
    log_interval = max(1, int(10 / FLAGS.num_cores))

    train_loader = torch.utils.data.DataLoader(
        datasets.MNIST(FLAGS.datadir, train=True, download=True,
                       transform=transforms.Compose([
                           transforms.ToTensor(),
                           transforms.Normalize((0.1307,), (0.3081,))
                       ])),
        batch_size=FLAGS.batch_size, shuffle=True, num_workers=FLAGS.num_workers)
    test_loader = torch.utils.data.DataLoader(
        datasets.MNIST(FLAGS.datadir, train=False, transform=transforms.Compose([
                           transforms.ToTensor(),
                           transforms.Normalize((0.1307,), (0.3081,))
                       ])),
        batch_size=FLAGS.batch_size, shuffle=True, num_workers=FLAGS.num_workers)

    model = MNIST()

    # Trace the model.                                                                                                                                                      devices = [':{}'.format(n) for n in range(0, FLAGS.num_cores)]
    xla_model = xmru.XlaModel(model, [FLAGS.batch_size, 1, 28, 28],
                              [FLAGS.batch_size], F.nll_loss,
                              num_cores=FLAGS.num_cores, devices=devices,
                              target_dtype=torch.int64)
    optimizer = optim.SGD(xla_model.parameters_list(), lr=lr, momentum=momentum)

    for epoch in range(0, FLAGS.num_epochs):
        xla_model.train(train_loader, optimizer, FLAGS.batch_size,
                        log_interval=log_interval)
accuracy = xla_model.test(test_loader,
                                  xmru.category_eval_fn(F.nll_loss),
                                  FLAGS.batch_size)
    return accuracy
```

**Comment by soumith** [comment]
Having an `XlaModel` that wraps `model` makes a lot of sense, considering the static sizing needed for XLA. The user should be aware of these semantic expectations. And, it gets rid of repeated boilerplate.

However, I dont think `XlaModel` should have `train` and `test` that'll contain the `optimizer` and effectively implement a training loop.
The reason is because this breaks down really quickly, even for basic stuff like a GAN, or a detection pipeline, or something like a slightly complex NMT model.

While being in control of the training loop gives us the ability to nicely control infeed, it makes it much more limited in what users can do, and I think we should hold off on that. This is actually the same reason PyTorch doesn't have an official trainer abstraction -- it's not obvious what to build yet.

The only big benefit of having the training loop be hidden is that (as you've stated in your comment) you can do device loops, implicit transfer overlaps etc.
I think there are cleaner user-aware APIs we can design to achieve the same, without blackboxing it.

**Comment by asuhan** [comment]
@dlibenzi Fusing the optimizer with the rest of the graph is harder than the loss. If we make it that the separate optimizer fuses with the rest of the graph, shouldn't the same approach work for the loss as well? Fundamentally, we can implement the full `Tensor` interface with lazy semantics (allowing auto-batching and fusion) and it'll support both cases.

@soumith If we figure out a way to overlap the transfer of next input batch to the TPU and running the model on the current batch, that'd go a long way towards minimizing the need for on-device loops. I have few ideas, but I'm sure there are a few I'm not aware of.

**Comment by dlibenzi** [comment]
> @dlibenzi Fusing the optimizer with the rest of the graph is harder than the loss. If we make it that the separate optimizer fuses with the rest of the graph, shouldn't the same approach work for the loss as well? Fundamentally, we can implement the full Tensor interface with lazy semantics (allowing auto-batching and fusion) and it'll support both cases.

Fusing the optimizer is pretty straightforward with the current XLA tensor infrastructure.
Essentially, at iteration 0, model parameters are just data on device.
When we come to the next iteration, we have, for example (Pj == Parameter 'j', 'i' = iteration number):

```
Pj[i+1] = optimizer_function(Pj[i], *optimizer_state)
```

The XLA tensors behind the model paramemters will have some XlaOp queued.
This means that the translator will have, instead of implicitly generating kParameter XlaOps, will have to get XLATensors and call the XlaGraphNode::Generate() to create the XlaOp which is operand of the graph to be compiled.
Since (hopefully) the optimizer function does not change over time, after the first iteration, we should be getting cache hits from the compilation cache.

So if we do:

```
for ;;
    out = xla_model(...)
    xla_loss.backward()
    optimizer.step()
```

At the first iteration _xla_model()_ will create an XLA computation where the model parameters are just data on device.
Since _XlaModel_ has loss in the traced model, we will fuse forward and backward.
At the next iteration, due to optimizer.step(), the model parameters will be mini XLA graphs which have previous version of the model parameters data as leaf nodes.
Iteration 0:

```
xla_forward_backward(a0, ...) {
  use(a0);
  ...
  return a0, ...
}

a0, ... = xla_forward_backward(a0, ...)
a1 = a0 + b * c // Optimizer step
```

Iteration 1..N:

```
xla_forward_backward(a0, b, c, ...) {
  a1 = a0 + b * c
  use(a1);
  return a1, ... // Return a1 as data so we can sync previous optimizer step ops
}

a1, ... = xla_forward_backward(a1, b, c, ...)
a2 = a1 + b * c // Optimizer step
```

After that should be steady state. We essentially fuse the optimizer step in front of the next iteration.


> @soumith If we figure out a way to overlap the transfer of next input batch to the TPU and running the model on the current batch, that'd go a long way towards minimizing the need for on-device loops. I have few ideas, but I'm sure there are a few I'm not aware of.

The TF ClientSession supports concurrent Run() calls, so we could in theory do it.
At step N, execute N and send input N+1, in parallel. Needs to be seen how TF/XRT execution behaves on the other side.
Currently models take input data for that step though, so normal signatures might have to change.
Clearly if we end up with a full fusion, the ROI of _infeed_ lowers.

**Comment by soumith** [comment]
the discussion seems to have converged with @dlibenzi 's last comment indicating that "magic under the hood" can comfortably optimize certain patterns pretty well.

I'll address some earlier comments, and expand on some points.

> I agree that the simple train/test case does not solve all the ML problems, and there are some that needs the user to have total control of the training loop, but a large set of the ML models training fits that pattern.

No, this is actually an anti-pattern. We are targeting typical PyTorch users, who tend more towards research, rather than something like Estimator API. Every model they hit needs a weird tweak in API.

> Ideally we want to fuse forward+backward+optimizer_step all in a single XLA computation.

Only if we see non-trivial wins, which I doubt we will see, based on my previous discussions with @asuhan . I actually dont see why we should prematurely fuse everything into a single XLA computation, if the perf win is small to non-existent.
To give guidance, we should be willing to be 10% slower in exchange for a more easier/flexible API. PyTorch philosophy has little patience for last-mile benchmarks or speed in exchange for more flexible API.

A big point that might have been missed (that we really need to keep in mind) is that 100% of the computation might not be able to map to TPU, especially on day 1 (or even several months in) -- so part of the computation might run on another device, CPU (or maybe even GPU?). It can be worse, part of computation might be in numpy (ugh!). We are aiming for TPU support with setting some workloads that are representative of each area  -- convnets for image, video, detection, RNN/transformer for NLP -- we are saying that proxy workloads such as resnet training, maskrcnn etc. are representative of each of these areas, but getting coverage of these particular workloads wont mean we will have coverage of entire area. 

> With that in place, are we OK in having a Python level XlaModel class which accepts a loss function, and implements a call interface?

What is the exact need for it, I still dont get it. And the interface really quickly breaks down practically.

To give you two simple examples (very common patterns):

```python
for ;;;
    out1 = model_main(input)
    out2 = model_aux(out1)
    err = loss_fn(out2, target)
    err.backward()
    optimizer.step()
```

```python
for ;;;
    fake = G()
    real = input
    pred = D(real)
    err = loss_fn(pred, targets)
    err.backward()
    pred = D(fake)
    err = loss_fn(pred, -targets)
    err.backward()
    optimizer.step()
```

The 2nd example, we'll hit as soon as we do anything GAN. The 1st example, we dont even have to get out of image classification domain

**Comment by dlibenzi** [comment]
> No, this is actually an anti-pattern. We are targeting typical PyTorch users, who tend more towards research, rather than something like Estimator API. Every model they hit needs a weird tweak in API.

> Only if we see non-trivial wins, which I doubt we will see, based on my previous discussions with @asuhan . I actually dont see why we should prematurely fuse everything into a single XLA computation, if the perf win is small to non-existent.
To give guidance, we should be willing to be 10% slower in exchange for a more easier/flexible API. PyTorch philosophy has little patience for last-mile benchmarks or speed in exchange for more flexible API.

There are a few reasons in trying to offer a path to top performance (I underline this - because it is not the **only** path for the user - things must still work for the other use cases) with fusion.
I can give you a real number for fusion (fresh of yesterday). Fusing backward and forward gave 20% boost on resnet18/cifar10.
And this is with a non-trivial part of the step, being sending data to server (we are looking into overlapping).
Once that part slims down, we can expect the fusion gain to be in the 30% range. This is not only computation wins, but the extra RPCs required by a non fused path.
Performance is also not the only win for fusion. Fusion can be the difference between OOM and fitting a model.
This is even a more compelling reason. One thing is saying a model runs 20+% slower (which can be big if you care about performance), and much worse is that the model just do not run at all.

> What is the exact need for it, I still dont get it. And the interface really quickly breaks down practically.

What I think was missed in the proposal analysis, is that we would still give a more "normal" PyTorch interface.
Example, the loss_fn is default-none, so if the user does not specify it, we would just return a non wrapped model.

I think it is important to underline again that what we are trying to design is not a "This is the only way you use PyTorch/TPU" thing.
Or "If we do not fuse we cannot function" (this can actually be true, since for resnet50 for example, we OOM in TPU v2 if we do not fuse).
But I hope it is clear what I mean in general.
The point is to design an interface which, while being as close as possible to the normal PyTorch semantics, offers a path to top TPU performance.
As TPUs are not free, researchers/engineers will be looking at TPUs (vs GPUs and CPUs) mainly for performance, and I believe there should be a path for them to get there.

**Comment by dlibenzi** [comment]
> oh, and before I forget -- overlapping input transfer with computation seems like it's good to do.
> How about we leverage something like [pytorch/pytorch#14171](https://github.com/pytorch/pytorch/pull/14171) to just transfer to XLA before it even gets out of DataLoader into main process?

We currently have a Python iterator wrapper over a dataloader object, which transfers data in background and effectively hands out XLATensors to the caller instead of Torch tensors.
AFAICT the pin_memory() seems like a CUDA specific concept.
I believe in order for that to work we would have to hook into the Torch tensor API (this is what the dataloader hands out), and the memory pin API should have a device argument?
Things can be worked out, but to collect numbers the iterator wrapper should suffice for now.
The above is just another API area which, if we want TPUs to be a first class citizen of PyTorch (and use common APIs), we might have to change.

**Comment by dlibenzi** [comment]
Just to give an idea of how the "legacy" XlaModel interface can look like, here is an example using a state of the art linear function learning :)
Here are some justifications about the API, which we can discuss here.
When we create an XLA model, we need to get the sample inputs, since on the way back we need to get the parameters to be passed to the optimizer constructor.
Since in the legacy interface we do not pass the loss to the XLA interface, we are essentially cutting the potentially monolithic graph, just at the tip, and hence we need to run the XLA model backward() explicitly.
This is about it, as far as "weird" things going on, AFAICT.
Using the same interface, with supplemental constructor arguments, the user will be able to achieve further optimizations, if he decides to.


```
class AxPlusB(nn.Module):
    def __init__(self):
	super(AxPlusB, self).__init__()
        self.a = nn.Parameter(torch.randn([]))
        self.b = nn.Parameter(torch.randn([]))

    def forward(self, x):
	return self.a * x + self.b

class SquareLoss(nn.Module):
    def __init__(self):
        super(SquareLoss, self).__init__()

    def forward(self, x, y):
        x.requires_grad = True
        y.requires_grad = True
        return (x - y) * (x - y)

class TestAxPlusB(XlaTestCase):
    def test(self):
	A = 3.11
        B = 4.09
        model = AxPlusB()
	xla_model = xmru.XlaModel(model, [torch.randn([])])
        optimizer = optim.SGD(xla_model.parameters_list(), lr=0.1, momentum=0.5)
        square_loss = SquareLoss()
	loss = None
	for _ in range(0, 100):
            optimizer.zero_grad()
            x = torch.randn([])
            target = A * x + B
            y = xla_model(x)
            loss = square_loss(y[0], target)
            loss.backward()
            xla_model.backward(y)
            optimizer.step()
        self.assertEqualRel(loss, torch.Tensor([0.0]))
```

**Comment by soumith** [comment]
in your example, the loss is on the CPU, and it's not an XlaModel. It shouldn't be this case in the final state of things. The loss should also be an XlaModel.

It might not be so in the legacy case just because we didn't implement the loss lowering to Xla, but there's no other reason to be so.

```python
xla_model = xmru.XlaModel(model, [torch.randn([])])
optimizer = optim.SGD(xla_model.parameters_list(), lr=0.1, momentum=0.5)
square_loss = SquareLoss()
xla_loss = xmru.XlaModel(square_loss, output_size)
loss = None
for _ in range(0, 100):
            optimizer.zero_grad()
            x = torch.randn([])
            target = A * x + B
            y = xla_model(x)
            loss = xla_loss(y[0], target)
            loss.backward()
            optimizer.step()
```

For the evolved interface, we could do what PyTorch actually does, which naturally allows fusing forward + backward if you'd want i.e.:

```python
model = MyModel(...)
loss = MyLoss(...)
traced = xmru.trace(lambda(input, target): return loss(model(input), target), input=(sample_input, sample_target))

optimizer = optim.SGD(traced.parameters_list(), lr=0.1, momentum=0.5)

for input, target in train_loader:
    optimizer.zero_grad()
    loss = traced(input, target)
    loss.backward()
    optimizer.step()
```

By not defining the `loss` function, but letting the user actually give a `closure` or `lambda` that we simply trace, we cover all of the applications that I mentioned above (GAN training, detection, etc. etc.), while addressing your need of fusing forward and backward.

Such a thing actually is already defined in PyTorch, for 2nd order optimizers.
If needed for max perf, I'm not opposed to defining a subset of optimizers in the xla package that follow the closure pattern, which can fuse the entire training step for one mini-batch.

https://pytorch.org/docs/stable/optim.html#optimizer-step-closure

**Comment by soumith** [comment]
oh, and as mentioned offline, to complete the conversation publicly as well, we absolutely shouldn't leave 20% to 30% of perf on the table for any reason whatsoever -- sorry if my comments came off as such. I'd only expect a max of 10% to be traded off, that to for essential or significant value in exchange.
The perf improvement numbers you quoted above in https://github.com/pytorch/xla/issues/4#issuecomment-440327747 made me realize we both weren't arguing over closing the 10% gap but over much more

**Comment by dlibenzi** [comment]
> For the evolved interface, we could do what PyTorch actually does, which naturally allows fusing forward + backward if you'd want i.e.:
> 
> ```python
> model = MyModel(...)
> loss = MyLoss(...)
> traced = xmru.trace(lambda(input, target): return loss(model(input), target), input=(sample_input, sample_target))
> 
> optimizer = optim.SGD(traced.parameters_list(), lr=0.1, momentum=0.5)
> 
> for input, target in train_loader:
>     optimizer.zero_grad()
>     loss = traced(input, target)
>     loss.backward()
>     optimizer.step()
> ```
> 
> By not defining the `loss` function, but letting the user actually give a `closure` or `lambda` that we simply trace, we cover all of the applications that I mentioned above (GAN training, detection, etc. etc.), while addressing your need of fusing forward and backward.
> 
> Such a thing actually is already defined in PyTorch, for 2nd order optimizers.

The lambda is a good idea. I thought that the tracer wanted a nn.Module.
This is why the evolved interface gets an nn.Module from the user ... and hack it :)

https://github.com/pytorch/xla/blob/21e98d584ff71675c6017db0f9d138c67f0a144b/torch_xla_py/xla_model.py#L347

I feel though that accepting an nn.Module has a lower entry point WRT the existing PyTorch modules around, instead of turning them to lambdas.
In the evolved interface we accept a loss function/lambda, so that the traced result is the full network.
At that point, when we run the forward, we already have the full graph we can monolithically lower, and get the performance and memory utilization benefits.


> in your example, the loss is on the CPU, and it's not an XlaModel. It shouldn't be this case in the final state of things. The loss should also be an XlaModel.
> 
> It might not be so in the legacy case just because we didn't implement the loss lowering to Xla, but there's no other reason to be so.
> 
> ```python
> xla_model = xmru.XlaModel(model, [torch.randn([])])
> optimizer = optim.SGD(xla_model.parameters_list(), lr=0.1, momentum=0.5)
> square_loss = SquareLoss()
> xla_loss = xmru.XlaModel(square_loss, output_size)
> loss = None
> for _ in range(0, 100):
>             optimizer.zero_grad()
>             x = torch.randn([])
>             target = A * x + B
>             y = xla_model(x)
>             loss = xla_loss(y[0], target)
>             loss.backward()
>             optimizer.step()
> ```

I am not sure what the saving would be in doing that, unless the loss function is really heavy weight.
What are the advantages you see in doing so?

**Comment by soumith** [comment]
> I am not sure what the saving would be in doing that, unless the loss function is really heavy weight.
What are the advantages you see in doing so?

Well, the loss function being on eager CPU will just prevent forward-backward fusion, that's all I was saying.

> The lambda is a good idea. I thought that the tracer wanted a nn.Module.
This is why the evolved interface gets an nn.Module from the user ... and hack it :)


Ok let's lock-in on such an interface. Is this reasonable (example for classification task)?

```python

def fn(inputs):
    input, target = inputs
    loss = F.cross_entropy(model(input), target)
    return loss

trace = xmru.trace(fn, inputs)

for input, target in train_data:
    optimizer.zero_grad()
    loss = trace((input, target))
    loss.backward()
    if i % 10:
        print(loss.item())
```

`train_fn` can be an inline lambda as well, or any python callable for that matter.

It will handle complex training loops such as GAN loops quite well. [Exhibit 1](https://github.com/pytorch/examples/blob/master/dcgan/main.py#L197-L231)
